text/0569-tracked-list.md

```diff
@@ -0,0 +1,306 @@
+is also the motivation behind the name: `TrackedArray` would imply a tracked
+version of native arrays, where `TrackedList` implies that it is its own data
+structure, with potentially divergent behaviors.
+
+## Detailed design
+
+`TrackedList` will re-implement the array API 1-1 for the most part, including
+native iterables in environments where they are available (enabling native
+destructuring syntax). Any existing prototype and static methods on `Array` will
+also exist on `TrackedList`, and future methods that are added will be added as
+soon as possible, without the need for an RFC (unless it would require
+significant changes to the way `TrackedList` works).
+
+As such, the rest of the design portion of this RFC will cover the tracking
+dynamics of `TrackedList`, and the ways in which it diverges from native arrays.
+
+### Tracking Dynamics
+
+In general, any changes to a `TrackedList` will be propagated as changes to the
+_whole_ list, rather than to individual indexes within the list. This is for two
+main reasons:
+
+1. Most changes will result in downstream consumers having to reprocess the
+   entire list, in general.
+2. It is much more performant to track changes to the list as a whole rather
+   than to each individual item within the list.
+
+In the following example, for instance, the `firstFriend` getter would be
+invalidated every time a new friend is added, even though it only accesses the
+first item in the list:
+
+```js
+import { TrackedList } from '@glimmer/tracking';
+
+class Person {
+  friends = TrackedList.from(['Yehuda', 'Melanie', 'Ricardo']);
+
+  addFriend(newFriend) {
+    this.friends.push(newFriend);
+  }
+
+  get firstFriend() {
+    let [first] = this.friends;
+
+    return first;
+  }
+}
+```
+
+From a change tracking perspective, we can see that this is actually very
+similar to the immutable pattern described at the beginning of this RFC - we
+are essentially telling the tracking system that some changes have occured, and
+now `friends` is _effectively_ a new array that needs to be reprocessed.
+
+In general, methods that _access_ the list and don't mutate it will entangle
+the list in the autotracking context. Methods that _mutate_ the list will
+invalidate, but not entangle the list, preventing backflow issues when
+rendering.
+
+### `getAt` and `setAt`
+
+It will not be possible to access the list items directly using `[]` syntax,
+like on normal arrays, since this would require native proxies (and would incur
+performance penalties). `TrackedList` will have two extra methods instead for
+direct index access: `getAt` and `setAt`. These methods will operate similar to
+`objectAt` and `replace` on `EmberArray` today, allowing the user to pass a
+numeric index as the first value, and in the case of `setAt` to pass the value
+to set the index to as the second:
+
+```ts
+declare class TrackedList<T> {
+  getAt(index: number): T;
+  setAt(index: number, value: T): T;
+}
+```
+
+### `sort` and `reverse`
```

Comment:
Borrowing an idea from Swift, perhaps we should name these `sorted` and `reversed`, the difference of names implying the difference in outcome?